### DISCUSSION TITLE
[RFC] Logs storage for apps

### DISCUSSION THREAD

**Discussion by lkostrowski** [saleor/saleor#14433] [RFC]
Co-authors: @maarcingebala @korycins @przlada 

## Problem

Apps even if small, still require to set up critical infrastructure like logging.

In multitenant apps we can imagine 2 types of logs:
1. Server-side - available on the infrastructure level, can be sent and consumed eg via Datadog. This is responsibility of app developer. Users are not aware of them and can't access them. Their main goal is to make app observable and reliable
2. Client-side (not to confuse with frontend) - "domain" logs that should be available to staff users who need to observe app behavior.

This issue is about client-side logs specifically. 

To store logs, app must have some sort of persistence (db or external service). Once connected, app can display such logs on it's frontend. However this is not perfect. App can be down and logs can't be accessed. App is also limited what it can do with such logs and every app developer must set up database just for that (but Metadata can be used as a quick workaround).

I suggest we add API that will store logs in Saleor.

## Business use-case

- Transaction apps - require to monitor performed calculations. For example product prices sent from Saleor to Taxes app and result of calculations.
- Debugging during app configuration - app can send logs optionally, when "debug mode enabled"
- Monitoring errors - dashboard can display last error log
- General Apps Monitoring  - logs can be sent via webhook to other platforms, emails etc

---

## APIs

```graphql
# Can be a directly implemented in App object - implementation detail
interface AppWithLogs {
  logs(input: AppLogsCountableConnectionInput): AppLogsCountableConnection
  lastErrorLog: AppLog
}
```

```graphql
type CountableConnectionInput {
  // ... pagination fields
  filter: {
    severity: LogLevelEnum
  }
}

```

```graphql
type AppLog implements ObjectWithMetadata {
  createdAt: Date;
  level: LogLevelEnum
  message: String!
  metadata: MetadataItem # Can be single item instead a list?
}
```

```graphql
enum LogLevelEnum {
  INFO, ERROR, WARNING
}
```

```graphql
mutation appLogsPush(input: [{
  message: String!;
  metadata: MetadataItem # Can it be a single item instead a list?
  level: LogLevelEnum
}]) {}
```

### Log structure

- createdAt - timestamp when log was created
- level - severity
- message - string message, log summary, user friendly message
- metadata - inherited from ObjectWithMetadata. Log can be enriched with additional payload

**Questions:**
**Q:** If metadata is used, it will set a list of metadata keys. For log its quite verbose and introduces additional sub-collection which is unnecessary, because logs is already quite atomic. App must decide what key to be set. Can Saleor alter behavior here to directly attach a JSON or force a key (to be eg `payload`). Or leave it like it is? 

**A:** Payload is optional so its not a problem to have it unstructured shape. Metadata enforces top level array, but its a valid json too. So metadata fits fine and app can decide how to push data there 

### Number of logs to persist

Number of logs should be set in configuration. Assumption is that enterprise-grade software will likely forward logs to single source of logs like Datadog. Saleor should not replicate such services features.

Limit should be set per-severity. For example "Debug" logs should not wipe error logs.

**Questions:**
- What should be the default value for each severity level?
- Should limit be set per-app or per-env?

### Log metadata limit

**Questions** 
- Is there any specific limit of metadata size?

### Fetching logs by app

App will include a paginated collection of logs and last log extracted on WRITE operation

### Dashboard can display logs

Dashboard can use "app" query to display last logs in "manage app" view. This view can be a dedicated place where simple app monitoring will appear.

Dashboard can provide "a button" to fetch collection of logs on demand

Thanks to that feature even if app is down, user can still debug some operations

### `APP_LOGS_CREATED` event

Saleor can inform about created logs with an event.

This allows:
1. Connecting another app that will forward logs e.g. to pager duty or slack
2. Stream all logs to logs aggegators 
3. Saleor can one day have native streaming of logs to APM together with other events

One day subscription can include a filter input that will filter which events are sent.

Event should be limited with items number and throttled:
1. Every X seconds event will be sent with a collection of at least one element
2. If no elements are queued, event is not sent
3. Event contains a maximum of Y events that are included in payload.
4. If there are more events queued, event is sent more than once every Y seconds.

**Questions**
- What should be time throttle value?
- What should be a batch size?

## `appLogsPush` mutation

App can push log for its own without a permission.

Mutation allows sending one or more logs in bulk

### Log severity

Log should have a top level field with a severity. Thanks to that:
1. There can be a separate number of events per severity that is stored
2. Events can prioritize higher severity and postpone lower severity
3. Querying logs can include filtering by severity level

### Adding logs with webhooks response

Additionally, especially for synchronous webhooks, app response can include logs. Saleor can use response to extract log structure and append log.

Example response:

```
{
shipping_price_gross_amount: number;
shipping_price_net_amount: number;
shipping_tax_rate: number;
lines: Array<{
      total_gross_amount: number;
      total_net_amount: number;
      tax_rate: number;
}>;
logs: Log[]
}
```

The `logs` should be extracted and there will be no need to send extra mutation

## MVP

Feature should start small and be released with a Preview Mode

Stage 1:
- Mutations to create logs
- Storing logs in DB
- Queries to fetch logs

Stage 2:
- Event 

Stage 3:
TBD

# Other open questions
- 

## Relation to Events Log
https://github.com/saleor/saleor/pull/12441

Event log stores state changes of App object, like:
- app installed
- app updated
- app permission granted
- app disabled

It represents lifecycle of the app. It can be consumed by Dashboard/staff to see what happened to the App, but also other APIs can build on top of it to resolve some state.

App Logs are domain logs/events that happens inside the app. App can use Private Metadata or any DB to implement this. 
Implementing this in Saleor is a convenient way to improve DX and time to market of the apps.

## Relation to observability event

Observability event is technical and tracks network between services. It doesn't cover domain behavior. It can be useful for monitoring, SRE, devops and developers. Its not useful for staff members.

App logs target staff user to be recipient and allow to store important events of the app domain. Its not a network abstraction

**Comment by maarcingebala** [comment]
I like this idea; it doesn't sound like a big effort to add something immediately useful. I think it shouldn't be generic API but a dedicated one for apps, since we have clear requirements here.

**Comment by lkostrowski** [comment]
> I like this idea; it doesn't sound like a big effort to add something immediately useful. I think it shouldn't be generic API but a dedicated one for apps, since we have clear requirements here.

I think you are right, when I try to imagine any other entity using logs - I can't. We are having Events Logs which can be part of any entity, but its a different abstraction:
- Event logs - set by Saleor, its basically an event store with important state changes
- Logs storage - just a bin of logs that only app write

So I will update the RFC

**Comment by lkostrowski** [comment]
Spawning @patrys to review

**Comment by patrys** [comment]
Initial comments:

1. What is the use case for having both INFO and DEBUG? How are those two different?
2. What is the difference between WARNING and ERROR? How do we expect API users and staff to react to both?
3. Do we need a new interface for this or would this just be a new field on App?
4. Is `payload` essentially not `metadata`? We don't want logs to become a storage mechanism so I'd rather give log entries metadata than allow arbitrarily large objects to be attached.
5. Do we need a separate `{ appLogs { ... } }` query or can the same be achieved using `{ app { logs { ... } } }`?
6. I think `App.logs` needs to be a connection field with pagination.
7. I think `App.logs` needs filtering based on log level.
8. I think we need to denormalize information about the most recent error timestamp at the App level, so the UI (Dashboard) can visually indicate apps with recent failures without having to read through the logs of each installed app.

**Comment by lkostrowski** [comment]
>What is the use case for having both INFO and DEBUG? How are those two different?

In this case I was thinking about following: 
1. INFO - medium severity, stored with history of X entries
2. DEBUG - low severity, Saleor can set different strategy on it, for example allow more items, but purge them every N minutes. DEBUG can be used to track app behavior during configuration, to understand what is happening. Once its set up, the "debug checkbox" can be disabled to reduce the traffic.

>What is the difference between WARNING and ERROR? How do we expect API users and staff to react to both?

1. They can have different strategies, eg warning can be treated like INFO but its just semantically different, error will persist more events
2. App that listens on `APP_LOGS_CREATED` can filter by this level to decide how to route these logs. E.g. error - pager duty, warning - email

We can drop this, but likely app will do this anyway by putting this information to payload/metadata. But DX will be worse

>Do we need a new interface for this or would this just be a new field on App?

From RFC perspective its an interface which app can implement, but I believe app can implement this directly / add to App class

>Is payload essentially not metadata? We don't want logs to become a storage mechanism so I'd rather give log entries metadata than allow arbitrarily large objects to be attached.

Thats actually a good idea, we don't have to reinvent the wheel. So I believe it will be `AppLog implements ObjectWithMetadata`?

The only issue I see is that root metadata structure will be an array, not an object and every time key must be known. Unless we use this indirectly (payload => map to key: "payload" etc)

I guess we need to verify the size of metadata -> will revisit this idea

>Do we need a separate { appLogs { ... } } query or can the same be achieved using { app { logs { ... } } }?

I don't remember why I used top level query. I think you are right, will update

>I think App.logs needs to be a connection field with pagination.

Yea, lets do it. Looks like I assumed nested logs in app is a simple list, while dedicated query is a paginated one. I will unify

>I think App.logs needs filtering based on log level.

Definitely, and one day also on subscription level to stream only certain level

>I think we need to denormalize information about the most recent error timestamp at the App level, so the UI (Dashboard) can visually indicate apps with recent failures without having to read through the logs of each installed

WDYT?

```graphql
app {
  logs{...}
  errorLogs: logs(severity: ERROR, last: 1) {...}
}

```

**Comment by patrys** [comment]
>DEBUG - low severity, Saleor can set different strategy on it, for example allow more items, but purge them every N minutes. DEBUG can be used to track app behavior during configuration, to understand what is happening. Once its set up, the "debug checkbox" can be disabled to reduce the traffic.

How is that different from telling the app to no longer log debug information at all? If the app keeps sending us DEBUG logs and Saleor is configured to ignore them, we are still wasting resources.

**Comment by patrys** [comment]
>WDYT?
>
>```graphql
>app {
>  logs{...}
>  errorLogs: logs(severity: ERROR, last: 1) {...}
>}
>```

That's expensive to calculate as it required a filtered join between tables, where something like `App.lastError` could be directly fetched from the database and only updated on write.

**Comment by korycins** [comment]
> I think we need to denormalize information about the most recent error timestamp at the App level, so the UI (Dashboard) can visually indicate apps with recent failures without having to read through the logs of each installed

And what should we do in case when staff user solved the error? We need a mutation to mark it as resolved on Saleor side.
Without marking it as resolved, he will still see the message on the dashboard side (and probably after X time he will start to ignore it assuming that we still display the old error).

**Comment by lkostrowski** [comment]
>And what should we do in case when staff user solved the error? We need a mutation to mark it as resolved on Saleor side.
Without marking it as resolved, he will still see the message on the dashboard side (and probably after X time he will start to ignore it assuming that we still display the old error).

I don't think it should work like this.

Logs is immutable event. Its not *state*.

We have another API in progress, Events Logs. This API is to compute state.

So it can work like this:
1. App breaks. It logs error and disables itself with some error message
2. User fixes the app and enables it

The history will show:
1. Historical logs
2. Computed state (app is enabled and working as a latest state event computed)

We can decide in the frontend if we want to show "last error log" always, or if state condition is met

Now I'd keep it simple and wait for Event logs

**Comment by korycins** [comment]
> so the UI (Dashboard) can visually indicate apps with recent failures 

---

> I don't think it should work like this.

> Logs is immutable event. Its not state.

---
And we don't modify any event. Your logs are still there, but if we want to display last error, we need to have a way to "clear" the lastError field
. 
Take into account this:
Assuming that we store logs for the last 2 weeks, and you have at least one error per 1-2 weeks. 
You will always see the `lastError` on the dashboard side. It will be an annoying thing - you already solved the problem on app side, but the dashboard still shows the `lastError`.  Each staff user contacts you because she/he sees the problem and asks you to solve it, but you already solved it. Each day you need to check if the `lastError` displayed on the dashboard has the same `createdAt` date as the previous one.

**Comment by lkostrowski** [comment]
I think it matters how we display this.

We can show error indicator with "error in last 30 minutes", otherwise don't - user will have to open a "logs tab" to see this historical logs

**Comment by korycins** [comment]
Yeap, that sounds good!

**Comment by patrys** [comment]
We can also store "last dismissed" timestamp as app metadata if you want to be able to hide errors until they reoccur.

**Comment by lkostrowski** [comment]
@patrys can we consider this RFC as "refined" from your point of view? There are few open questions left, but maybe we can answer them during the implementation (like size limits)

**Comment by patrys** [comment]
From my end, the only question is what to do with the `metadata` comments. I don't think I understand what the comment about "a single item" means.

**Comment by lkostrowski** [comment]
> From my end, the only question is what to do with the `metadata` comments. I don't think I understand what the comment about "a single item" means.

ObjectWithMetadata that I expect will be used, requires entity to have a collection of metadata with keys and values.

This adds extra nesting and verbosity. App will have to decide what metadata key to write and read, but for log context likely a single item with a json is enough

However, We can also treat `MetadataItem` as a single json entry, so instead

```json
[
  {
    "key": "payload",
    "value": "{ ... stringified json here }"
  }
]
```

we can assume this:

```json
[
  {
    "key": "orderId",
    "value": "12345"
  }
]

```

Since payload is an additional field, I think we can leave it unstructured like metadata

I updated the content of the rfc

**Comment by patrys** [comment]
In that case, I'd prefer it implement the same interface as all other objects for consistency.
